fix(http): correct Tower service usage and preserve backpressure

BrynCooke opened the pull request:
## Summary

Fixed incorrect Tower service usage in the HTTP client that was breaking backpressure flow control. The service was consuming itself on each call instead of being reusable, which violates Tower's service patterns and could lead to connection issues.

## Problem

The HTTP client service was incorrectly using Tower's `call()` method which consumes the service, making it unusable for subsequent requests. This breaks Tower's backpressure mechanism that services rely on for proper flow control. Without proper service reuse, each request created unnecessary overhead and could lead to connection problems when making requests to subgraph services.

## Solution

Implemented proper Tower service patterns with concurrent caching:

- Added service caching in `HttpClientServiceFactory` using `DashMap` for lock-free concurrent access
- Used `ServiceBuilder` with buffered cloning to maintain proper backpressure while enabling service reuse
- Changed subgraph service from `call()` to `oneshot()` for correct Tower service consumption
- Added tests that cover cache hits, misses, and service isolation by name

## Technical Details

The implementation ensures Tower services follow their intended patterns:

- **Backpressure Preservation**: Services are cached and reused rather than consumed, maintaining flow control through the service chain
- **Concurrent Access**: `DashMap` provides thread-safe caching without blocking or lock contention
- **Service Lifecycle**: Buffered cloning allows multiple concurrent requests while preserving proper Tower service semantics

## Review Strategy

- **Service Lifecycle**: Examine `HttpClientServiceFactory::create()` method for proper caching and service creation logic
- **Tower Patterns**: Verify the change from `call()` to `oneshot()` in subgraph service interaction follows Tower conventions
- **Backpressure Flow**: Check that buffered services properly propagate `poll_ready` signals through the service chain
- **API Compatibility**: Confirm constructor changes don't break existing `HttpClientServiceFactory` usage patterns
- **Test Coverage**: Review `test_service_caching()` for adequate validation of caching behavior

<!-- start metadata -->
---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- [ ] Metrics and logs are added[^3] and documented
- Tests added and passing[^4]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: A lot of (if not most) features benefit from built-in observability and `debug`-level logs. Please read [this guidance](https://github.com/apollographql/router/blob/dev/dev-docs/metrics.md#adding-new-metrics) on metrics best-practices.
[^4]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

Commit:
fix(http): correct Tower service usage and preserve backpressure

The HTTP client service was incorrectly using call() which consumes the service
and breaks Tower's backpressure mechanism. This implements proper service
cloning to maintain backpressure while enabling service reuse.

Technical changes:
- Added memoized HashMap in HttpClientServiceFactory to cache services by name
- Implemented Arc<RwLock> wrapper for thread-safe service caching
- Used ServiceBuilder with buffered cloning for proper Tower service patterns
- Changed subgraph service from call() to oneshot() for correct consumption

Review focus:
- Examine HttpClientServiceFactory::create() method for proper service lifecycle
- Verify the RwLock usage in the cache read/write blocks doesn't introduce deadlocks
- Check that the buffered service in create() properly propagates poll_ready backpressure
- Confirm the memoized field addition doesn't break existing constructor usage

Commit:
changeset

Commit:
lint

Commit:
changelog

Commit:
perf(http): replace RwLock with DashMap for service caching

The RwLock-based service cache could potentially cause writer starvation
under high contention scenarios where many threads are reading cached
services while new services need to be inserted.

Switched to DashMap which provides lock-free concurrent access patterns
and eliminates the possibility of writer starvation. This simplifies
the caching logic and improves performance under concurrent load.

Technical changes:
- Replaced Arc<RwLock<HashMap>> with Arc<DashMap> for service cache
- Simplified cache access logic by removing explicit lock management
- Removed Arc wrapper from MemoizedService as DashMap handles interior mutability

Review focus:
- Verify the cache access pattern in HttpClientServiceFactory::create() method
- Confirm DashMap semantics match the expected caching behavior for concurrent access

garypen commented on .changesets/maint_refactor_http_service_clone.md:
```diff
@@ -0,0 +1,7 @@
+### HTTP client service now allows backpressure ([PR #7694](https://github.com/apollographql/router/pull/7694))
+
+This internal refactoring caches http_client services rather than recreated them on every call. 
```
```suggestion
This internal refactoring caches http_client services rather than recreating them on every call. 
```

Commit:
test(http): add service caching verification test

Added test to verify that HttpClientServiceFactory properly caches services
by name and reuses them for subsequent requests. This ensures the DashMap
implementation correctly handles concurrent access patterns and prevents
service recreation overhead.

Technical changes:
- Added test-only methods cache_len() and has_cached_service() for introspection
- Implemented comprehensive test covering cache hits, misses, and isolation
- Added explanatory comment about Buffer usage for thread safety

Review focus:
- Verify test coverage adequately validates caching behavior in HttpClientServiceFactory::create()
- Check that test-only methods don't expose implementation details inappropriately

garypen commented on apollo-router/src/services/http.rs:
```diff
@@ -61,17 +76,29 @@ impl HttpClientServiceFactory {
-            })
+        // Check if we already have a memoized service for this name
+        if let Some(service) = self.cache.get(name) {
+            service.deref().clone().boxed()
```
```suggestion
            service.clone().boxed()
```

garypen commented on apollo-router/src/services/http.rs:
```diff
@@ -1,8 +1,14 @@
 #![allow(dead_code)]
+
+use std::ops::Deref;
```
```suggestion
```

garypen reviewed (commented):
Looks good. The `deref()` is redundant.

BrynCooke committed:
Update .changesets/maint_refactor_http_service_clone.md

Co-authored-by: Gary Pennington <gary@apollographql.com>

BrynCooke committed:
Update .changesets/maint_refactor_http_service_clone.md

BrynCooke committed:
Update apollo-router/src/services/http.rs

Co-authored-by: Gary Pennington <gary@apollographql.com>

BrynCooke committed:
Update apollo-router/src/services/http.rs

Co-authored-by: Gary Pennington <gary@apollographql.com>

BrynCooke commented on .changesets/maint_refactor_http_service_clone.md:
```diff
@@ -0,0 +1,7 @@
+
+This internal refactoring caches http_client services rather than recreated them on every call. 
+This has no impact on user functionality but is retained in the changelog so that other teams
+can be alerted to this change in behavior.
```
[c03052f](https://github.com/apollographql/router/pull/7694/commits/c03052fc19500727f8ac3dad9c08e75e3cb87f75)

goto-bus-stop commented:
Oh, I didn't realise http_client_service is only for private plugins, maybe we don't need to mention it then 🙈

Commit:
Merge branch 'dev' into refactor/http-service-clone

Commit:
Merge branch 'dev' into refactor/http-service-clone

Commit:
Merge branch 'dev' into refactor/http-service-clone

Commit:
After performance investigations RwLock is the winner

A bunch of small changes to the existing PR which aim to optimise
performance:

 - less boxing
 - RwLock<HashMap> seems to outperform DashMap

BrynCooke committed:
After performance investigations RwLock is the winner (#7733)

BrynCooke committed:
Merge branch 'dev' into refactor/http-service-clone
